include/triton/core/tritonbackend.h

```diff
@@ -1356,6 +1356,51 @@ TRITONBACKEND_ModelInstanceReportStatistics(
     const uint64_t compute_start_ns, const uint64_t compute_end_ns,
     const uint64_t exec_end_ns);
 
+/// Record statistics for a decoupled inference response.
+///
+/// All timestamps should be reported in nanonseconds and collected using
+/// std::chrono::steady_clock::now().time_since_epoch() or the equivalent.
+///
+/// Call this function after calling TRITONBACKEND_ResponseSend, and pass the
+/// same send_flags and error object passed to the TRITONBACKEND_ResponseSend.
+///
+/// For consistency of measurement across different backends, the
+/// timestamps should be collected at the following points during
+/// TRITONBACKEND_ModelInstanceExecute.
+///
+///   TRITONBACKEND_ModelInstanceExecute()
+///     < start of this response >
+///     CAPTURE TIMESPACE (response_start)
+///     < generate this response >
+///     CAPTURE TIMESPACE (compute_output_start)
+///     < allocate output buffers and extract output tensors, including copying
+///       the tensors to/from GPU if necessary >
+///     CAPTURE TIMESPACE (response_end)
+///     < end of this response >
+///     return
+///
+/// \param instance The model instance.
+/// \param response_factory The response factory associated with the inference
+/// request.
+/// \param response_start Timestamp for the start of execution for this
+/// response.
+/// \param compute_output_start Timestamp for the start of extracting output
+/// tensors for this response. Set this to 0 for reporting empty response.
+/// \param response_end Timestamp for the end of extracting output tensors for
+/// this response.
+/// \param send_flags Flags associated with the response. \see
+/// TRITONBACKEND_ResponseSend
+/// \param error The TRITONSERVER_Error to send if the response is an error, or
+/// nullptr if the response is successful. \see TRITONBACKEND_ResponseSend
+/// \return a TRITONSERVER_Error indicating success or failure.
+TRITONBACKEND_DECLSPEC TRITONSERVER_Error*
+TRITONBACKEND_ModelInstanceReportResponseStatistics(
+    TRITONBACKEND_ModelInstance* instance,
+    TRITONBACKEND_ResponseFactory* response_factory,
+    const uint64_t response_start, const uint64_t compute_output_start,
+    const uint64_t response_end, const uint32_t send_flags,
```

Comment:
Not a main reviewer here, but @Tabrizian @kthui @oandreeva-nv do you think we should use another Opaque object here for stats/info if we need to add more fields to this in the future? Currently appending another field would require a new API.

If we sent a `TRITONBACKEND_ResponseStats` object or something, we can append to it?

This could also be the object to denote if the request was cancelled for phase 2 , etc

Reply:
Also for backends that implement inflight batching themselves, or perhaps some other logic I'm not currently thinking of, backends may maintain their own queues. This could also be a place to report how long a request was "queued" from the backend's perspective, rather than triton core scheduler perspective.

Just food for thought - may not make total sense yet.

Reply:
Yes, me and @kthui had an offline discussion that in future there may be a need for new API wich handles `key` and does not use `response_factory`. I think some opaque object could help

Reply:
I think you are on the right track @rmccorm4. My thought ultimately is to allow the backend to specify a "key" which is used to group/aggregate the "same" response statistics across different requests. I will leave the decision to @Tabrizian if we want to switch to / implement the "key" now.

Reply:
@rmccorm4 what is the use-case you have in mind? Is it possible for the user to use custom metrics for anything that doesn't fit into this structure?

> The distinction between request-specific and response-specific is a little fuzzy for me atm.

For decoupled models, currently there is no way to access the latency breakdown for individual responses. These statistics help the user to get a better idea about the latency breakdown of individual responses.

Reply:
I think custom metrics *can* be used for things that fall outside of the current stats, but was just curious if we had any ideas of other stats that may be useful to add to the "core" set in the near-mid term. If so, we would have to add a new C API for new stats if we didn't use some opaque object.

For example, if the use of iterative scheduling / IFB handled in Triton core had more adoption - would we want some notion of per-step stats for rescheduled requests? I think this case might be analogous to the "empty/flag-only" intermediate responses, where we don't actually produce a response but want to somehow signify that we have done some work in between.

Reply:
> I think this case might be analogous to the "empty/flag-only" intermediate responses, where we don't actually produce a response but want to somehow signify that we have done some work in between.

I don't think response stats is the right place for this. If you are sending a response, you can use response stats.